packages/go/cypher/analyzer/measure.go

```diff
@@ -0,0 +1,271 @@
+		s.nodeLookupKinds[typedReference.Symbol] = s.nodeLookupKinds[typedReference.Symbol].Add(node.Kinds...)
+	}
+
+	return nil
+}
+
+func (s *ComplexityMeasure) onMerge(_ *model.WalkStack, node *model.Merge) error {
+	// Let's add 1 per merge action
+	s.Weight += weight1 * int64(len(node.MergeActions))
+
+	return nil
+}
+
+func (s *ComplexityMeasure) onNodePattern(_ *model.WalkStack, node *model.NodePattern) error {
+	if node.Binding == nil {
+		if len(node.Kinds) == 0 {
+			// Unlabeled, unbound nodes will incur a lookup of all nodes in the graph
+			s.Weight += weight2
+		}
+	} else if nodePatternBinding, typeOK := node.Binding.(*model.Variable); !typeOK {
+		return fmt.Errorf("expected variable for node pattern binding but got: %T", node.Binding)
+	} else {
+		nodeLookupKinds, hasBinding := s.nodeLookupKinds[nodePatternBinding.Symbol]
+
+		if !hasBinding {
+			nodeLookupKinds = node.Kinds
+		} else {
+			nodeLookupKinds = nodeLookupKinds.Add(node.Kinds...)
+		}
+
+		// Track this node pattern to see if any subsequent expressions will narrow its kind matchers
+		s.nodeLookupKinds[nodePatternBinding.Symbol] = nodeLookupKinds
+	}
+
+	if node.Properties != nil {
+		s.hasPatternProperties = true
+	}
+
+	return nil
+}
+
+func (s *ComplexityMeasure) onPartialComparison(_ *model.WalkStack, node *model.PartialComparison) error {
+	switch node.Operator {
+	case model.OperatorRegexMatch:
+		// Regular expression matching incurs a weight since it can be far more involved than any of the other
+		// string operators
+		s.Weight += weight1
+	}
+
+	return nil
+}
+
+func (s *ComplexityMeasure) onPatternPart(_ *model.WalkStack, node *model.PatternPart) error {
+	// All pattern parts incur a compounding weight
+	s.numPatterns += 1
+	s.Weight += s.numPatterns
+
+	if node.ShortestPathPattern {
+		// Rendering the shortest path, while cheaper than rendering all shortest paths, still could incur a large
+		// search cost
+		s.Weight += weight1
+	}
+
+	if node.AllShortestPathsPattern {
+		// Rendering all shortest paths could result in a large search
+		s.Weight += weight2
+	}
+
+	return nil
+}
+
+func (s *ComplexityMeasure) onProjection(_ *model.WalkStack, node *model.Projection) error {
+	// We want to capture the cost of additional inline projections so ignore the first projection
+	s.Weight += s.numProjections
+	s.numProjections += 1
+
+	if node.Distinct {
+		// Distinct incurs a weight since it will change how the projection is materialized
+		s.Weight += weight1
+	}
+
+	if node.Limit != nil {
+		s.hasLimit = true
+	}
+
+	return nil
+}
+
+func (s *ComplexityMeasure) onQuantifier(_ *model.WalkStack, _ *model.Quantifier) error {
+	// Quantifier expressions may increase the size of an inline projection to apply its contained filter and should
+	// be weighted
+	s.Weight += weight1
+	return nil
+}
+
+func (s *ComplexityMeasure) onRelationshipPattern(_ *model.WalkStack, node *model.RelationshipPattern) error {
```

Comment:
I realize most of this code is just being moved from the analyzer.go file but I was wondering why we name the `*model.RelationshipPattern` parameter as `node`. Similarly for other param names in these methods. I might expect it to be named `relationshipPattern` like in `liftRelationshipPatternCriteria`. 

The reason I was curious was because I see that `match (n) return n` will be gated by setting it to be over the max complexity allowed. I often search for a specific edge in the cypher search, e.g., `match p= ()-[:DCSync]->() return p` which does not have any "node" constraints but does have a relationship constraint. Would DCSync then be included in the `ComplexityMeasure`'s `nodeLookupKinds`? Or will this query for returning these DCSync edges be disallowed because it has no "node" lookup kinds?

Reply:
I'm not entirely sure why node was chosen as the name for these parameters, other than maybe because they are nodes in the parser tree? Regardless, we did a decent amount of work to ensure that _any_ constraint would be sufficient to avoid that max complexity call onExit. That includes relationship labels, node labels, property filters, any where clause, or a limit.

Reply:
Yea nodes in the parser tree was something I was considering. Okay, that gives me peace of mind to know that edge matching like I wrote would still be allowed. Thanks for the explanation!

Reply:
If you look at the JSON tests, you should see a few tests that validate that an edge label is sufficient as well. We tried to add any tests that seemed to be missing to ensure we'd catch any known edge cases

Reply:
```
{
    "name": "Query for the properties of an edge using keys()",
    "type": "string_match",
    "details": {
        "query": "match ()-[e:EDGE_OF_INTEREST]->() return keys(e)",
        "complexity": 6
    }
},
```